File: Structures/Structures/Tree/BinaryTree.swift
//
//  BinaryTree.swift
//  Structures
//
//  Created by MinKyeongTae on 2020/02/03.
//  Copyright © 2020 MinKyeongTae. All rights reserved.
//

// MARK: BinaryTree Structure Implementation by Swift

// MARK: - Basic BinaryTree Structure 코드구현 복습

/*
 class BinaryTreeNode<T> {
 var value: T
 var leftChild: BinaryTreeNode?
 var rightChild: BinaryTreeNode?
 init(_ value: T, _ leftChild: BinaryTreeNode? = nil, _ rightChild: BinaryTreeNode? = nil) {
     self.value = value
     self.leftChild = leftChild
     self.rightChild = rightChild
 }

 func traverseInOrder(_ visit: (T) -> ()) {
     self.leftChild?.traverseInOrder(visit)
     visit(self.value)
     self.rightChild?.traverseInOrder(visit)
 }

 func traversePreOrder(_ visit: (T) -> ()) {
     visit(self.value)
     self.leftChild?.traversePreOrder(visit)
     self.rightChild?.traversePreOrder(visit)
 }

 func traversePostOrder(_ visit: (T) -> ()) {
     self.leftChild?.traversePostOrder(visit)
     self.rightChild?.traversePostOrder(visit)
     visit(self.value)
 }
 }

 // MARK: - Below Nodes

 //                1
 //        2               3
 //    4       5       6       7

 // InOrder : 4, 2, 5, 1, 6, 3, 7
 // PreOrder : 1, 2, 4, 5, 3, 6, 7
 // PostOrder : 4, 5, 2, 6, 7, 3, 1

 var leaf7 = BinaryTreeNode(7)
 var leaf6 = BinaryTreeNode(6)
 var leaf5 = BinaryTreeNode(5)
 var leaf4 = BinaryTreeNode(4)
 var node3 = BinaryTreeNode(3, leaf6, leaf7)
 var node2 = BinaryTreeNode(2, leaf4, leaf5)
 var rootNode = BinaryTreeNode(1, node2, node3)
 rootNode.traverseInOrder { print("\($0) ", terminator: "")} // 4,2,5,1,6,3,7
 print()
 rootNode.traversePreOrder { print("\($0) ", terminator: "")} // 1,2,4,5,3,6,7
 print()
 rootNode.traversePostOrder { print("\($0) ", terminator: "")} // 4,5,2,6,7,3,1
 */

// MARK: - BinaryTree 기본구조 Swift 구현

/*
 import Foundation

 class BinaryTreeNode<T> {
 var value: T
 var leftChild: BinaryTreeNode?
 var rightChild: BinaryTreeNode?
 init(_ value: T, _ leftChild: BinaryTreeNode? = nil, _ rightChild: BinaryTreeNode? = nil) {
     self.value = value
     self.leftChild = leftChild
     self.rightChild = rightChild
 }

 func traverseInOrder(_ visit: (T) -> ()) {
     leftChild?.traverseInOrder(visit)
     visit(self.value)
     rightChild?.traverseInOrder(visit)
 }

 func traversePreOrder(_ visit: (T) -> ()) {
     visit(self.value)
     leftChild?.traversePreOrder(visit)
     rightChild?.traversePreOrder(visit)
 }

 func traversePostOrder(_ visit: (T) -> ()) {
     leftChild?.traversePostOrder(visit)
     rightChild?.traversePostOrder(visit)
     visit(self.value)
 }
 }

 let five = BinaryTreeNode<Int>(5)
 let four = BinaryTreeNode<Int>(4)
 let three = BinaryTreeNode<Int>(3)
 let two = BinaryTreeNode<Int>(2, five, nil)
 let one = BinaryTreeNode<Int>(1, three, four)
 let rootTree = BinaryTreeNode<Int>(0, one, two)

 // 3 1 4 0 5 2
 print("traverseInOrder Rusult ▼")
 rootTree.traverseInOrder { print($0, terminator: "->") }
 print()

 // 0 1 3 4 2 5
 print("traversePreOrder Rusult ▼")
 rootTree.traversePreOrder { print($0, terminator: "->") }
 print()

 // 3 4 1 5 2 0
 print("traversePostOrder Rusult ▼")
 rootTree.traversePostOrder { print($0, terminator: "->")}

 */

File: Structures/Structures/Tree/BinaryTree.cpp
//
//  BinaryTree.cpp
//  Structures
//
//  Created by MinKyeongTae on 2020/02/02.
//  Copyright © 2020 MinKyeongTae. All rights reserved.
//

// MARK: Binary Tree

#if 0
#include <iostream>
using namespace std;
template <typename T> // 다형성을 위해 타입은 템플릿을 사용,
class Node {
private:
    T value;
    Node *left;
    Node *right;
    Node *root;
public:
    Node() : value(0), left(nullptr), root(nullptr) {};
    Node(T _value) : value(_value), right(nullptr), root(nullptr) {};
    Node(T _value, Node* _left, Node* _right) : value(_value), left(_left), right(_right), root(nullptr) {
        // 만약 왼쪽 오른쪽 노드가 존재한다면, 현재 노드는 부모노드이다.
        if(nullptr != _left)
            _left->root = this;
        if(nullptr != _right)
            _right->root = this;
    };
    
    ~Node() {};
    
    void setLeft(Node* node) { this->left = node; }
    void setRight(Node* node) { this->right = node; }
    Node* getLeft() { return left; }
    Node* getRight() { return right; }
    T getValue() { return value; }
};

// 재귀를 사용한 전위 / 중위 / 후위 노드 출력
template <typename T>
void printPreorder(Node<T>* node) // 전위
{
    if(node == nullptr) return;
    cout << node->getValue() << " ";
    printPreorder(node->getLeft());
    printPreorder(node->getRight());
}

template <typename T>
void printInorder(Node<T>* node) // 중위
{
    if(node == nullptr) return;
    printInorder(node->getLeft());
    cout << node->getValue() << " ";
    printInorder(node->getRight());
}

template <typename T>
void printPostOrder(Node<T>* node) { // 후위
    if(node == nullptr) return;
    printPostOrder(node->getLeft());
    printPostOrder(node->getRight());
    cout << node->getValue() << " ";
}

int main() {
    ios_base :: sync_with_stdio(0); cin.tie(0); cout.tie(0);
    
    // TREE STATUS
        
    //            3
    //        1       2
    //      4   5   6
    
    // inorder : 4 1 5 3 6 2
    // preorder : 3 1 4 5 2 6
    // postorder : 4 5 1 6 2 3
    
    Node<int>* node1 = new Node<int>(1);
    Node<int>* node2 = new Node<int>(2);
    Node<int>* node3 = new Node<int>(3, node1, node2);
    Node<int>* node4 = new Node<int>(4);
    Node<int>* node5 = new Node<int>(5);
    Node<int>* node6 = new Node<int>(6);

    node1->setLeft(node4);
    node1->setRight(node5);
    node2->setLeft(node6);
    
    cout << "전위 결과 : ";
    printPreorder(node3);
    puts("");
    cout << "중위 결과 : ";
    printInorder(node3);
    puts("");
    cout << "후위 결과 : ";
    printPostOrder(node3);
    puts("");
    return 0;
}
#endif
